### ISSUE TITLE
Kubernetes Batch System

### ISSUE THREAD

**Issue by cricketsloan** [DataBiosphere/toil#2717]
Since our lab is probably going to deploy a Kubernetes cluster, and since we have access to the Pacific Research Platform which uses Kubernetes, we ought to have Kubernetes support in Toil.

The easy part is probably the batch system. There's another harder question of providing Toil's Docker functions in an environment where the Docker daemon is not allowed to start and all Docker access must be mediated by Kubernetes, but that's probably a separate issue to work on after this one.

┆Issue is synchronized with this [Jira Story](https://ucsc-cgl.atlassian.net/browse/TOIL-385)
┆Issue Number: TOIL-385

**Comment by adamnovak** [comment]
@jeffrey856 Here's an example script I put together to talk to Kubernetes, using the credentials and namespace in your `~/.kube/config`.

It might work out of the box with Minikube; for the PRP/Optiputer system I had to create a service account and set up my Kubernetes to use it:

```
kubectl apply -f - <<EOF
apiVersion: v1
kind: ServiceAccount
metadata:
  name: adamnovak-service
EOF

kubectl apply -f - <<EOF
apiVersion: v1
kind: Secret
metadata:
  name: adamnovak-service-secret
  annotations:
    kubernetes.io/service-account.name: adamnovak-service
type: kubernetes.io/service-account-token
EOF

kubectl apply -f - <<EOF
apiVersion: rbac.authorization.k8s.io/v1
kind: Role
metadata:
  namespace: ucsc-cgl
  name: job-runner
rules:
- apiGroups: [""]
  resources: ["pods"]
  verbs: ["get", "list", "watch"]
- apiGroups: ["batch"]
  resources: ["jobs"]
  verbs: ["get", "list", "watch", "create", "update", "patch", "delete"]
EOF

kubectl create rolebinding adamnovak-service-user \
  --role=job-runner \
  --serviceaccount=ucsc-cgl:adamnovak-service \
  --namespace=ucsc-cgl

kubectl describe secrets/adamnovak-service-secret
```

And then I put the key in my `~/.kube/config`:

```
apiVersion: v1
clusters:
- cluster:
    certificate-authority-data: LS0tLS1CRUdJTiBDRVJUSUZJQ0FURS0tLS0tCk1JSUN5RENDQWJDZ0F3SUJBZ0lCQURBTkJna3Foa2lHOXcwQkFRc0ZBREFWTVJNd0VRWURWUVFERXdwcmRXSmwKY201bGRHVnpNQjRYRFRFNE1ESXhOVEExTWpNek9Wb1hEVEk0TURJeE16QTFNak16T1Zvd0ZURVRNQkVHQTFVRQpBeE1LYTNWaVpYSnVaWFJsY3pDQ0FTSXdEUVlKS29aSWh2Y05BUUVCQlFBRGdnRVBBRENDQVFvQ2dnRUJBTTZxCm5tNUVSb3FGVHNzWWtJNUExNmZCVTJNbDN6SUpWZnN1dXY3aEpzNk1VMUlCdURuQTJQM0xkaGlpdXFpb0NNamQKRGdwcG92WTVnUVFOZXp1UVo0OGJiQmhac2dWZ1gwK3BUWUlPVjV6ZUo1bGdaclpzWHJsSWZJTHREc3BpcEo0YQpadTRiNnVuVmptSk1LeU4vTHdaRWlvU2d3MXdLRkhlcmF2czA4NHZzQ1F3R2NZNGh4bnhSNzJjdkpjdFFHeHVyCmRnUVNBd2V3U2oycXZWaGJHRzhWU3M3Mms1OU00QlIvMDhrMUxnVmF0alZtYUcvcG9VSnBETTkycVYzTnRwNjcKQUdaRnBNNjhYNHpwSU15NDlDMTNpY0hsZzRGVlNLblNiMUFvWlB0UWUrdGpNMWxTT2FPOWk3U2NxczFzWEdxYwpDazJLUFo3OEUzQUxzMVlpK0swQ0F3RUFBYU1qTUNFd0RnWURWUjBQQVFIL0JBUURBZ0trTUE4R0ExVWRFd0VCCi93UUZNQU1CQWY4d0RRWUpLb1pJaHZjTkFRRUxCUUFEZ2dFQkFIZ3hmRmwzVkk3eW9Sbzc3M25OOGpEMDhhT1cKQmwxZldORktZQTJhY3ZTZXFsTnk4WGdkaDNMQ3JGVUI0N24vTFlaUVhEeklEVnU5M0RZQ21EbzM1dFZiMXRDMwpUL1BNd202RnpkTW1NY2M3a210dkR2YXhXNDBWbFU5WGptZ214QnhieXd3VTZLUGtCUzVZTjRGZUpEakUvSTVWCll3Ui9HeXJrS2JiL0NYbURrUDNON2orQWEvS1FXcDVPMnA3aUVibFBsTTFrazJ3Q2VVS1NuYTR0MjgvbE1CWEcKRmhNRXA2S3pIYmVXcnJKdGdjbTBLZkRkT25xZm4zbU1rYmpDbk0wV2FiZUN3eEpIWXZnbXR4RkVaTGNhODhGZAppTEF0Rms4UElUTVVXaTg2allXQmFmQkFJRW5mckJaY05mMitVMkF5S0dNN2lBeUMrZWFyQzBBOEJZOD0KLS0tLS1FTkQgQ0VSVElGSUNBVEUtLS0tLQo=
    server: https://67.58.53.147:6443
  name: nautilus
contexts:
- context:
    cluster: nautilus
    namespace: ucsc-cgl
    user: adamnovak-service
  name: nautilus
current-context: nautilus
kind: Config
preferences: {}
users:
- name: adamnovak-service
  user:
    token: PASTE_TOKEN_HERE
```

Here's the script:

```
#!/usr/bin/env python3

"""
run.py: run a command via Kubernetes
"""

import kubernetes
import uuid
import time



def find_kubernetes():
    """
    Set up global Kubernetes context.

    Work out what namespace to run in.

    Returns the namespace of the active context.
    """

    # Load ~/.kube/config
    kubernetes.config.load_kube_config()

    # Find all contexts and the active context.
    # The active context gets us our namespace.
    contexts, active_context = kubernetes.config.list_kube_config_contexts()
    if not contexts:
        raise RuntimeError("No Kubernetes contexts available in ~/.kube/config")

    print("Active context: {}".format(active_context))

    return active_context.get('context', {}).get('namespace', 'default')



def run_command(command, image, namespace, cores=1.0, mem_bytes=1 * 1024 * 1024 * 1024, disk_bytes=10 * 1024 * 1024 * 1024, basename="job-"):
    """
    Run the given command (a list of strings) in the given Docker image
    specifier. Run the job under the given Kubernetes namespace name.
    
    Uses the given float number of CPU cores and the given number of bytes of
    memory and (ephemeral) disk to run the job.

    Returns the Kubernetes V1Job, which has a .metadata.name field that can be
    used to get its status.
    """

    # The Kubernetes API makes sense only in terms of the YAML format. Objects
    # represent sections of the YAML files. Except from our point of view, all
    # the internal nodes in the YAML structure are named and typed.

    # For docs, start at the root of the job hierarchy:
    # https://github.com/kubernetes-client/python/blob/master/kubernetes/docs/V1Job.md

    # Make a definition for the container's resource requirements
    requirements_dict = {'cpu': cores, 'memory': mem_bytes, 'ephemeral-storage': disk_bytes}
    resources = kubernetes.client.V1ResourceRequirements(limits=requirements_dict, requests=requirements_dict)
    print("Made resource requirements: {}".format(resources))
    # Make a volume to provision disk
    volume_name = 'tmp'
    volume_source = kubernetes.client.V1EmptyDirVolumeSource()
    volume = kubernetes.client.V1Volume(name=volume_name, empty_dir=volume_source)
    print("Made volume: {}".format(volume))
    # Make a mount for the volume
    volume_mount = kubernetes.client.V1VolumeMount(mount_path='/tmp', name=volume_name)
    print("Made volume mount: {}".format(volume_mount))
    # Make a container definition
    container = kubernetes.client.V1Container(command=command, image=image, name="runner-container", resources=resources, volume_mounts=[volume_mount])
    print("Made container: {}".format(container))
    # Wrap the container in a spec
    pod_spec = kubernetes.client.V1PodSpec(containers=[container], volumes=[volume], restart_policy="Never")
    print("Made pod spec: {}".format(pod_spec))
    # Wrap the spec in a template
    template = kubernetes.client.V1PodTemplateSpec(spec=pod_spec)
    print("Made template: {}".format(template))
    # Make another spec for the job, asking to run the template with backoff
    job_spec = kubernetes.client.V1JobSpec(template=template, backoff_limit=1)
    print("Made job spec: {}".format(job_spec))
    # Make metadata to tag the job with info.
    # We use generate_name to ensure a unique name
    metadata = kubernetes.client.V1ObjectMeta(generate_name=basename)
    print("Made metadata: {}".format(metadata))
    # And make the actual job
    job = kubernetes.client.V1Job(spec=job_spec, metadata=metadata, api_version="batch/v1", kind="Job")
    print("Made job: {}".format(job))
    
    # Get a versioned Kubernetes batch API
    api = kubernetes.client.BatchV1Api()

    # Make the job
    # TODO: How do we know what namespaces are available to use?
    launched = api.create_namespaced_job(namespace, job)

    print('Launched job: {}'.format(launched))

    return launched

def to_count(val):
    """
    Turn None into 0, or pass through a number.
    """

    if val is None:
        return 0
    return val

def wait_on_job(job):
    """
    Wait on the given Kubernetes V1Job.

    Returns the job's status when it is done.
    """

    # Get a versioned Kubernetes batch API
    api = kubernetes.client.BatchV1Api()

    # We could read_namespaced_job_status if we knew we had jobs/status
    # permission. But it's possible to be able to make  and list jobs but not
    # have that.

    # TODO: We should watch for our job instead of constantly polling it.
    # See https://stackoverflow.com/a/56260527 for an example of using
    # streaming info from Kubernetes.

    # Work out how to filter for the job
    field_selector = 'metadata.name=' + job.metadata.name

    # Get the list of jobs matching the selector
    found = api.list_namespaced_job(job.metadata.namespace, field_selector=field_selector, watch=False)

    print("Found {} jobs".format(len(found.items)))
    
    while True:
        current_job = api.read_namespaced_job(job.metadata.name, job.metadata.namespace)

        print("Job status: {}".format(current_job.status))

        if to_count(current_job.status.succeeded) + to_count(current_job.status.failed) == 1:
            # Job has succeeded or failed
            break

        time.sleep(1)

    # Clean up the job
    api.delete_namespaced_job(job.metadata.name, job.metadata.namespace)
        
if __name__ == "__main__":
    namespace = find_kubernetes()
    running_job = run_command(["sleep", "30"], "ubuntu:18.04", namespace)
    wait_on_job(running_job)
```
